### PR TITLE
Incorporate more 3.16 features

### LINKED ISSUE #448: Support for Monikers
New Monikers feature: https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#textDocument_moniker

More info on protocol here: https://github.com/microsoft/vscode-languageserver-node/blob/master/protocol/src/common/protocol.moniker.proposed.ts

### LINKED ISSUE #455: Support Trace notifications
New in LSP 3.16, there is now [LogTrace Notification](https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#logTrace) and [SetTrace Notification](https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#setTrace).

### LINKED ISSUE #457: Removing Semantic Highlighting
> Given [this](https://github.com/Microsoft/vscode-languageserver-node/pull/367), should we remove or deprecate semantic highlights as semantic tokens replace it?

_Originally posted by @ericdallo in https://github.com/eclipse/lsp4j/pull/446#issuecomment-683374512_

This question was never addressed, so I'm creating a separate issue for it. Now that Semantic Tokens are officially in the LSP 3.16 proposal and implemented in LSP4J, the Semantic Highlighting operations should be removed or deprecated. What is the view on beta/experimental features? Should they just be removed or should they be deprecated first even though they were never finalized?

### LINKED ISSUE #469: Support for `textDocument/onTypeRename`
LSP 3.16 provides a new `onTypeRename` support https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#textDocument_onTypeRename that we would like to consume in our XML Language Server to manage mirror cursor capability (when you update a start tag element, it will update the end tag element).

It should be really nice that lsp4j provides this support. Thanks!

### PR DESCRIPTION
**Pull request by KamasamaK**
Many changes have been made for the 3.16 release of LSP, so I've collected a lot of them.

### Added

- Change annotations for `WorkspaceEdit` (via `ResourceOperation`) and `AnnotatedTextEdit`
- Linked editing (Fixes #469)
- Opening and showing a document in the client user interface
- Refresh support for code lenses and semantic tokens
- Provide the client's locale in the initialize call
- Preserve additional attributes on `MessageActionItem`
- Capability to signal whether the client normalizes line endings
- Insert text mode on `CompletionItem`s
- Insert/replace for `CompletionItem` text edits
- Default behavior for prepare rename
- Overlapping and multiline support for semantic tokens
- Monikers (Fixes #448)
- Trace notifications (Fixes #455)
- Regular expression engine support
- Markdown parser support
- Progress support for rename, color provider, folding range provider, and document range formatting
- Missing properties on `ApplyWorkspaceEditResponse`

### Removed

- Non-standard feature Coloring is removed as it was already deprecated in favor of Semantic Highlighting which is itself deprecated in favor of Semantic Tokens (See #457)

### Compatibility issues

- Changed some types from `Number` to `Integer`
- `ProgressParams` missing nullability checks
- `CompletionItem.textEdit` now an `Either`

### CODE REVIEW

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/InsertTextMode.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+public enum InsertTextMode {
+	
+	/**
+	 * The insertion or replace strings is taken as it is. If the
+	 * value is multi line the lines below the cursor will be
+	 * inserted using the indentation defined in the string value.
+	 * The client will not apply any kind of adjustments to the
+	 * string.
+	 */
+  AsIs(1),
```
Comment (szarnekow) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/InsertTextMode.java:32]: Nitpick: Indentation looks off.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/MonikerKind.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * The moniker kind.
+ * 
+ * Since 3.16.0
+ */
+@Beta
+public final class MonikerKind {
+  private MonikerKind() {
+  }
+
+  /**
```
Comment (szarnekow) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/MonikerKind.java:27]: Mixed tabs and spaces?

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 3]
```diff
@@ -1154,34 +1286,6 @@ class FoldingRangeCapabilities extends DynamicRegistrationCapabilities {
...
- * <b>Note:</b> the <a href=
- * "https://github.com/Microsoft/vscode-languageserver-node/pull/367">{@code textDocument/semanticHighlighting}
- * language feature</a> is not yet part of the official LSP specification.
- *
- * @deprecated Use {@code SemanticTokens} API instead.
- */
-@Beta
-@Deprecated
-@JsonRpcData
-class SemanticHighlightingCapabilities {
```
Comment (szarnekow) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:1170]: In theory the removal would require a major version bump on LSP4J.
Not sure about the policy here.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 4]
```diff
@@ -3376,20 +3820,22 @@ class ProgressParams {
 	/**
 	 * The progress token provided by the client or server.
 	 */
-	Either<String, Number> token
+	@NonNull
+	Either<String, Integer> token
```
Comment (szarnekow) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:3824]: Not sure about this. The LSP spec says it's a number which is 64bit. Integer would only be 32bit.

**Review by szarnekow — COMMENTED** [review]
Thank you for the effort, @KamasamaK .
I'm not a maintainer of lsp4j but a consumer thus I try to follow the changes that are made. I've to say this is a huge merge request. Unfortunately it mixes different changes (all towards 3.16, but still). Would it by chance be feasible to split that up into smaller digestible chunks?

I'm through with the review yet but wanted to post the first small batch of comments and trigger this discussion.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend (hunk 4 above)
Comment (KamasamaK) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:3824]: `ProgressToken` type is defined as `integer | string` in LSP 3.16 -- https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#progress. `integer` type defined as "in the range of -2^31 to 2^31 - 1" -- https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#number

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend (hunk 3 above)
Comment (KamasamaK) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:1170]: I asked in #457 and one of the other admins said to remove them after the 0.10.0 release.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend (hunk 4 above)
Comment (szarnekow) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:3824]: I see. I was coming from https://microsoft.github.io/language-server-protocol/specification#progress. Sad.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend (hunk 3 above)
Comment (szarnekow) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:1170]: It's just a question of versioning the lsp4j plugin. Does this removal lead to a major version increment or not.

Comment (KamasamaK) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend:1170]: I don't deal with versioning or release management for LSP4J, but my understanding is that anything with the `@Beta` annotation can be changed (or removed) at any time. Breaking compatibility between minor versions seems to be acceptable in this project based on previous releases. I don't know if the versioning scheme used for this project is clearly defined anywhere, but I know one of the most popular versioning schemes SemVer says

https://semver.org/#spec-item-4
> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.
<br>
If anyone has strong feelings about it staying a bit longer, I would be fine with that if adoption of semantic tokens is not widespread enough.

**Comment by nixel2007** [comment]
Any chance to add `DocumentSymbolOptions` and `label` field?
p.s. thanks for your work!

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/InsertTextMode.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+import com.google.common.annotations.Beta;
+
+/**
+ * How whitespace and indentation is handled during completion
+ * item insertion.
+ *
+ * Since 3.16.0
+ */
+@Beta
```
Comment (nixel2007) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/InsertTextMode.java:22]: Why beta?

Comment (KamasamaK) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/InsertTextMode.java:22]: This PR is 2 weeks old which is before it was finalized. You will see a lot of that annotation. I will have to remove all of those now.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend (hunk 3 above)
Comment (KamasamaK) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: I've decided to create a separate PR for the removal of semantic highlighting.

**Pull request merged**